Add telemetry to the console host to report platform and version

JamesWTruher opened the pull request:
Produce simple telemetry via Application Insights for the PowerShell Core console host.

This is limited to the console host and is not meant as generalized telemetry code for PowerShell Core. It will capture the GitCommitID and Platform Information when the console host starts. It enables opting out of sending telemetry. This does not include the changes required in setup to opt out of telemetry, which will be in a later PR.

JamesWTruher committed:
Add console host Application Insights telemetry

add application insights reference
Create telemetry class
We will upload GitCommitId and OSVersion information
This is just applicable to the console host, if we want to make it more
general, we can refactor at that time

JamesWTruher committed:
Create a semaphore file to enable or disable console host telemetry

We need a way to allow the user to opt out of telemetry and with the config
RFC still in flight, we don't have a centralized way to do that. This implements
a semaphore file which when present will produce telemetry sends
from the console host when it starts up. In order to disable telemetry, simply
delete the file. Setup will need to be changed to handle this.

PaulHigin commented on build.psm1:
```diff
@@ -160,6 +160,9 @@ function Start-PSBuild {
     git --git-dir="$PSScriptRoot/.git" describe --dirty --abbrev=60 > "$psscriptroot/powershell.version"
 
+    # create the telemetry flag file
+    $null = new-item -force -type file "$psscriptroot/DELETE_ME_TO_DISABLE_CONSOLEHOST_TELEMETRY"
```
>$psscriptroot [](start = 40, length = 13)

This will create the file in the directory where Start-PSBuild is being run.  But don't we want this location to be the output publish path?

JamesWTruher replied:
The placement of the file in the actual publish directory is managed by putting it in the *.csproj files (same as powershell.version is done), but the file has to exist.

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -0,0 +1,63 @@
+        {
+            // if the semaphore file exists, send the telemetry
+            string assemblyPath = Path.GetDirectoryName(typeof(PSVersionInfo).GetTypeInfo().Assembly.Location);
+            string telemetrySemaphoreFilePath = Path.Combine(assemblyPath, TelemetrySemaphoreFilename);
```
This path shouldn't change per session so it seems they can also be static variables initialized on first use (for cases other then start up telemetry reporting).  You could use a static constructor or static assignment.

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -0,0 +1,63 @@
+        // Telemetry client to be reused when we start sending more telemetry
+        private static TelemetryClient telemetryClient = null;
+        // TODO: Set this to false for release
+        private static bool developerMode = true;
```
we should set this to false now.  if you need to test it, change it in your branch

SteveL-MSFT replied:
also rename to _developerMode

JamesWTruher replied:
fixed

---
In reply to: [112784357](https://github.com/PowerShell/PowerShell/pull/3620#discussion_r112784357) [](ancestors = 112784357)

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -0,0 +1,63 @@
+        public const string TelemetrySemaphoreFilename = "DELETE_ME_TO_DISABLE_CONSOLEHOST_TELEMETRY";
+
+        // Telemetry client to be reused when we start sending more telemetry
+        private static TelemetryClient telemetryClient = null;
```
convention has been to use leading underscore `_telemetryClient`

JamesWTruher replied:
fixed

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -0,0 +1,63 @@
+            var properties = new Dictionary<string, string>();
+            properties.Add("GitCommitID", PSVersionInfo.GitCommitId);
+            properties.Add("OSVersionInfo", Environment.OSVersion.VersionString);
+            SendTelemetry("PSCoreStartup", properties);
```
shouldn't this be ConsoleHostStartup?

JamesWTruher replied:
sure - fixed

---
In reply to: [112784414](https://github.com/PowerShell/PowerShell/pull/3620#discussion_r112784414) [](ancestors = 112784414)

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -0,0 +1,63 @@
+            // if the semaphore file exists, send the telemetry
+            string assemblyPath = Path.GetDirectoryName(typeof(PSVersionInfo).GetTypeInfo().Assembly.Location);
+            string telemetrySemaphoreFilePath = Path.Combine(assemblyPath, TelemetrySemaphoreFilename);
+            if ( File.Exists(telemetrySemaphoreFilePath ) )
```
>if ( File.Exists(telemetrySemaphoreFilePath ) ) [](start = 12, length = 47)

Some of these method calls can throw (Access Denied, ArgumentException) so I think we need to wrap this in a try/catch to allow PowerShell to continue to run on telemetry send fail.

lzybkr replied:
We have a fast version of `File.Exists` that we should probably use here for faster startup - it's `Utils.NativeFileExists`.

JamesWTruher replied:
excellent, thanks
fixed

---
In reply to: [112802981](https://github.com/PowerShell/PowerShell/pull/3620#discussion_r112802981) [](ancestors = 112802981)

PaulHigin commented:
:clock1:

PaulHigin reviewed (changes requested):
Please see my review comments

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:
```diff
@@ -257,6 +257,9 @@ internal sealed partial class ConsoleHost
                     }
 
+                    // Send startup telemetry for ConsoleHost startup
+                    ApplicationInsightsTelemetry.SendPSCoreStartupTelemetry();
```
Can you measure the impact on startup?
I see that ApplicationInsights claims calls are non-blocking and telemetry is batched and sent on another thread, so there may be little or no impact.
I have a feeling we'll want to cross-gen the ApplicationInsights assembly though - JIT time is often noticeable.

JamesWTruher replied:
i did some checking earlier (it was about 30ms), but will validate again since the code has changed to using a semaphore file. I will include that in the comments of the next push 

---
In reply to: [112802910](https://github.com/PowerShell/PowerShell/pull/3620#discussion_r112802910) [](ancestors = 112802910)

lzybkr replied:
30ms is still quite a bit - it would be ~10% of total startup if we were matching performance of Windows PowerShell.

It would be good to know how much of that time is in JIT, see  https://github.com/PowerShell/PowerShell/issues/3638 for some relative comparisons for time spent in JIT on other assemblies.

If the time is mostly in JIT (say > 95% of the time), then these changes are fine, and we can address cross-gen in the issue I've opened. If the time is not JIT time, then I think it should move to the background - for an example of where we do something similar in the console, see https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs#L52

JamesWTruher replied:
@lzybkr initially I was also calculating a checksum of the SMA assembly which was removed, and now there's less than 2ms increase when sending telemetry. (see comments below).

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -0,0 +1,63 @@
+        private static void SendTelemetry(string eventName, Dictionary<string,string>payload)
+        {
+            // if the semaphore file exists, send the telemetry
+            string assemblyPath = Path.GetDirectoryName(typeof(PSVersionInfo).GetTypeInfo().Assembly.Location);
```
To get PSHOME, we normally use this code:
```C#
                var psHome = Utils.GetApplicationBase(Utils.DefaultPowerShellShellID);
```

JamesWTruher replied:
excellent, will fix

---
In reply to: [112802942](https://github.com/PowerShell/PowerShell/pull/3620#discussion_r112802942) [](ancestors = 112802942)

JamesWTruher committed:
update telemetry class to use a static path for the semaphore file

Also, protect the telemetry sending from exception. A failure to create the telemetry client should not be fatal
to the shell starting.

JamesWTruher committed:
changes to use standard PS apis to get PSHOME and check for file existance

have git ignore the telemetry semaphore file

SteveL-MSFT reviewed (approved):
LGTM

JamesWTruher commented:
@lzybkr from a performance perspective, the cost is quite low. I did the following test with the semaphore file present and then not present:
`measure-command { 1..60 | %{ ./powershell.exe -c "exit" }`
the average difference was about +2ms per invocation when the semaphore file was present. I'm happy to use other methodologies, just let me know what you would like to see.

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -0,0 +1,78 @@
+        /// The name of the file by when present in $PSHOME will enable telemetry.
+        /// If this file is not present, no telemetry will be sent.
+        /// </summary>
+        public const string TelemetrySemaphoreFilename = "DELETE_ME_TO_DISABLE_CONSOLEHOST_TELEMETRY";
```
Does this need to be public? It's recommended to not have a public const field because this will make changing this const value a breaking change to other assemblies that reference `ConsoleHost.dll` and use this const field.

More specifically, if another assembly uses it, this string will be embedded in that assembly instead of actually referencing `ConsoleHost.dll` and retrieving the value at runtime. So in future, say we have to change this string value, the existing assemblies that depend on this public const field will continue to use the old values even after it's changed in `ConsoleHost.dll`.

JamesWTruher replied:
it doesn't need to be public. Do you think the TelemetrySemaphoreFilePath should also be private?

---
In reply to: [113082894](https://github.com/PowerShell/PowerShell/pull/3620#discussion_r113082894) [](ancestors = 113082894)

daxian-dbw replied:
I think it's better to be private because it's more like an implementation detail.

JamesWTruher replied:
i will make it private - the path as well?

---
In reply to: [113084013](https://github.com/PowerShell/PowerShell/pull/3620#discussion_r113084013) [](ancestors = 113084013)

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -0,0 +1,78 @@
+        /// </summary>
+        public static string TelemetrySemaphoreFilePath = Path.Combine(
+            Utils.GetApplicationBase(Utils.DefaultPowerShellShellID),
+            TelemetrySemaphoreFilename);
```
This also seems to be an implementation detail that shouldn't be exposed.

JamesWTruher replied:
changed to private

---
In reply to: [113083044](https://github.com/PowerShell/PowerShell/pull/3620#discussion_r113083044) [](ancestors = 113083044)

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -0,0 +1,78 @@
+    /// <summary>
+    /// send up telemetry for startup
+    /// </summary>
+    public class ApplicationInsightsTelemetry
```
It seems this could be declared as a static class, as it's not supposed to be instantiated.

JamesWTruher replied:
changed

---
In reply to: [113083648](https://github.com/PowerShell/PowerShell/pull/3620#discussion_r113083648) [](ancestors = 113083648)

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -0,0 +1,78 @@
+                if (Utils.NativeFileExists(TelemetrySemaphoreFilePath))
+                {
+                    TelemetryConfiguration.Active.InstrumentationKey = _psCoreTelemetryKey;
+                    TelemetryConfiguration.Active.TelemetryChannel.DeveloperMode = _developerMode;
```
Is this configuration a one-time thing, or do we have to do this configuration every time calling `SendTelemetry`? If it's a one-time thing, we probably can put it in the static constructor of this class.

JamesWTruher replied:
right now, it's a one-time thing. I'll create a static constructor and move this inside

---
In reply to: [113083840](https://github.com/PowerShell/PowerShell/pull/3620#discussion_r113083840) [](ancestors = 113083840)

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -0,0 +1,78 @@
+        {
+            var properties = new Dictionary<string, string>();
+            properties.Add("GitCommitID", PSVersionInfo.GitCommitId);
+            properties.Add("OSVersionInfo", Environment.OSVersion.VersionString);
```
`Environment.OSVersion.VersionString` doesn't give much useful information on different distro of Linux, for example, it show `Unix 4.8.0.41` on Ubuntu 16.04. How about use `[System.Runtime.InteropServices.RuntimeInformation]::OSDescription`?
```
PS /> [System.Runtime.InteropServices.RuntimeInformation]::OSDescription                            
Linux 4.8.0-41-generic #44~16.04.1-Ubuntu SMP Fri Mar 3 17:11:16 UTC 2017
```

JamesWTruher replied:
Yes, it should be the description, but I'm waiting for this to be available as part of PSVersionInfo, this is ok for now.

---
In reply to: [113084599](https://github.com/PowerShell/PowerShell/pull/3620#discussion_r113084599) [](ancestors = 113084599)

SteveL-MSFT replied:
I'd suggest changing to OSDescription now and using PSVersionTable is a change we can make later

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -0,0 +1,78 @@
+        /// <summary>
+        /// Create the startup payload and send it up
+        /// </summary>
+        public static void SendPSCoreStartupTelemetry()
```
Do we want to expose this API? By looking at the name, it's for a very specific scenario -- sending telemetry at PS startup time, and thus doesn't seem to be needed to made public.

JamesWTruher replied:
I'll make it internal

---
In reply to: [113084933](https://github.com/PowerShell/PowerShell/pull/3620#discussion_r113084933) [](ancestors = 113084933)

daxian-dbw replied:
Sounds good. In that case, the class `ApplicationInsightsTelemetry` should also be made internal.

PaulHigin reviewed (approved):
LGTM

JamesWTruher committed:
move class to internal and report on OSDescription rather than OS version

create static constructor and initialize active telemetry in it
change members to private rather than public

daxian-dbw commented:
Minor comment: Some of the using directives seem not necessary, and it's better to have them cleaned up.
